clang/lib/CodeGen/Targets/SPIR.cpp

```diff
@@ -64,6 +66,27 @@ void CommonSPIRABIInfo::setCCs() {
   RuntimeCC = llvm::CallingConv::SPIR_FUNC;
 }
 
+ABIArgInfo SPIRVABIInfo::classifyReturnType(QualType RetTy) const {
+  if (getTarget().getTriple().getVendor() != llvm::Triple::AMD)
+    return DefaultABIInfo::classifyReturnType(RetTy);
+  if (!isAggregateTypeForABI(RetTy) || getRecordArgABI(RetTy, getCXXABI()))
+    return DefaultABIInfo::classifyReturnType(RetTy);
+
+  if (const RecordType *RT = RetTy->getAs<RecordType>()) {
+    const RecordDecl *RD = RT->getDecl();
+    if (RD->hasFlexibleArrayMember())
+      return DefaultABIInfo::classifyReturnType(RetTy);
+  }
+
+  // TODO: The AMDGPU ABI is non-trivial to represent in SPIR-V; in order to
+  // avoid encoding various architecture specific bits here we return everything
+  // as direct to retain type info for things like aggregates, for later perusal
+  // when translating back to LLVM/lowering in the BE. This is also why we
+  // disable flattening as the outcomes can mismatch between SPIR-V and AMDGPU.
+  // This will be revisited / optimised in the future.
```

Comment:
I don't understand the difficulty representing it. SPIR/SPIRV just made the a mistake by using byval for anything (where they just took the default clang happens to emit). There's no real reason SPIR-V needs to continue using byval in kernels for its IR, and can switch to byref for kernels? What is the underlying SPIR-V representation like?

Reply:
I think you misunderstand the core problem, which is orthogonal to LLVM: SPIR-V the IR specification does not [encode byref, only byval](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#_function_parameter_attribute). Whatever you might feel about what is right or wrong is somewhat unhelpful here, unless you are going to add an extension to that end? Trying to backdoor this is just a lot of pain.

Reply:
To wit, it's necessary pain, in the end, but we've not yet figured out the right way to do it, since there are a few moving pieces involved. Perhaps the solution will end up to add `byref` to the SPIR-V spec; in the meanwhile, this is a sufficient, if temporary and slightly icky, solution.

Reply:
byval doesn't really make sense on a kernel as there is no real caller, but depending on how SPIRV defined its byval attribute, maybe you can codegen LLVM byref into SPIRV byval?

Reply:
> byval doesn't really make sense on a kernel as there is no real caller, but depending on how SPIRV defined its byval attribute, maybe you can codegen LLVM byref into SPIRV byval?

This is a good suggestion, it's one of the things I'm looking at, but it's a bit messy to retain the "this is actually `byref`" (it is doable, but was quite intrusive with tools like the translator), and I admit to not fully debugging / validating it. I just realised that the "add an extension to SPIR-V" bit came off pretty passive-aggressive, I apologise - I think that might just be TheRightWayTM to deal with this, `byref` is common enough to justify it.